airbyte-oauth/src/main/java/io/airbyte/oauth/google/GoogleOAuthFlow.java

```diff
@@ -53,59 +55,53 @@
 
   private final ConfigRepository configRepository;
 
-  public GoogleOAuthFlow(ConfigRepository configRepository) {
-    this(configRepository, HttpClient.newBuilder().version(Version.HTTP_1_1).build());
+  public GoogleOAuthFlow(ConfigRepository configRepository, String scope) {
+    this(configRepository, scope, HttpClient.newBuilder().version(Version.HTTP_1_1).build());
   }
 
   @VisibleForTesting
-  GoogleOAuthFlow(ConfigRepository configRepository, HttpClient httpClient) {
+  GoogleOAuthFlow(ConfigRepository configRepository, String scope, HttpClient httpClient) {
     this.configRepository = configRepository;
     this.httpClient = httpClient;
+    this.scope = scope;
+    this.googleQueryParameters = List.of(
+        String.format("scope=%s", this.scope),
+        "access_type=offline",
+        "include_granted_scopes=true",
+        "response_type=code",
+        "prompt=consent");
   }
 
   @Override
   public String getSourceConsentUrl(UUID workspaceId, UUID sourceDefinitionId, String redirectUrl) throws IOException, ConfigNotFoundException {
     final JsonNode oAuthParamConfig = getSourceOAuthParamConfig(workspaceId, sourceDefinitionId);
-    if (oAuthParamConfig.has("client_id")) {
-      final String clientId = oAuthParamConfig.get("client_id").asText();
-      return getConsentUrl(sourceDefinitionId, clientId, redirectUrl);
-    } else {
-      throw new IOException("Undefined parameter 'client_id' for Google OAuth Flow.");
-    }
+    return getConsentUrl(sourceDefinitionId, getClientIdUnsafe(oAuthParamConfig), redirectUrl);
   }
 
   @Override
   public String getDestinationConsentUrl(UUID workspaceId, UUID destinationDefinitionId, String redirectUrl)
       throws IOException, ConfigNotFoundException {
     final JsonNode oAuthParamConfig = getDestinationOAuthParamConfig(workspaceId, destinationDefinitionId);
-    if (oAuthParamConfig.has("client_id")) {
-      final String clientId = oAuthParamConfig.get("client_id").asText();
-      return getConsentUrl(destinationDefinitionId, clientId, redirectUrl);
-    } else {
-      throw new IOException("Undefined parameter 'client_id' for Google OAuth Flow.");
-    }
+    return getConsentUrl(destinationDefinitionId, getClientIdUnsafe(oAuthParamConfig), redirectUrl);
   }
 
-  private static String getConsentUrl(UUID definitionId, String clientId, String redirectUrl) {
-    final StringBuilder result = new StringBuilder(GOOGLE_ANALYTICS_CONSENT_URL)
+  private String getConsentUrl(UUID definitionId, String clientId, String redirectUrl) {
+    final StringBuilder result = new StringBuilder(consentUrl)
         .append("?");
-    for (String queryParameter : GOOGLE_QUERY_PARAMETERS) {
+    for (String queryParameter : googleQueryParameters) {
       result.append(queryParameter).append("&");
     }
     return result
+        // TODO state should be randomly generated, and the 2nd step of oauth should verify its value
+        // matches the initially generated state value
```

Comment:
In that case, the state value should be returned by the `getConsentUrl` and then, forwarded by the UI back to the `completeOAuthFlow` endpoint?

Reply:
maybe just passing the full "consentUrl" in the call from the UI? but that requires storing some state on that side.